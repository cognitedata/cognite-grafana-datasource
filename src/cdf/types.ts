export type Datapoints = Items<Datapoint>;
export type Events = Items<Event>;

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export interface TimeSeriesDatapoint {
  timestamp: number;
  value: string;
}

export interface TimeSeriesAggregateDatapoint {
  timestamp: number;
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface CogniteEvent {
  id: number;
  lastUpdatedTime: string;
  createdTime: string;
  externalId?: string;
  startTime?: number;
  endTime?: number;
  dataSetId?: number;
  description?: string;
  type?: string;
  subtype?: string;
  assetIds?: number[];
  source?: string;
  sourceId?: string;
  metadata?: { [s: string]: string };
}

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  activeAtTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export type EventsFilterTimeParams =
  | Pick<EventsFilterRequestParams, 'activeAtTime'>
  | Pick<EventsFilterRequestParams, 'startTime' | 'endTime'>;

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter?: Filter;
  advancedFilter?: any;
}

export interface Resource {
  id: number;
  externalId?: string;
  name?: string;
  description?: string;
}

export interface TimeSeriesResponseItem extends Resource {
  isString?: boolean;
  metadata?: Record<string, string>;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface Timestamp {
  timestamp: number;
}

export type Items<T = any> = {
  items: T[];
};

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export interface Range<T> {
  min?: T;
  max?: T;
}

export interface CogniteLabelsResponse {
  externalId: string;
  name: string;
  description: string;
  dataSetId: number;
  createdTime: number;
}
interface CogniteRelationshipAsset {
  createdTime?: number;
  lastUpdatedTime?: number;
  rootId?: number;
  aggregates?: {
    childCount: number;
    depth: number;
    path?: [
      {
        id: string | number;
      }
    ];
  };
  parentId?: number;
  parentExternalId?: string;
  externalId: string;
  name?: string;
  description?: string;
  dataSetId: number;
  metadata?: { [s: string]: string };
  source?: string;
  labels?: CogniteLabelsResponse[];
  id: number;
}
export interface CogniteRelationshipResponse {
  externalId: string;
  sourceExternalId: string;
  sourceType?: string;
  targetExternalId: string;
  targetType?: string;
  startTime?: number;
  endTime?: number;
  confidence?: number;
  dataSetId?: number;
  labels?: CogniteLabelsResponse[];
  createdTime?: number;
  lastUpdatedTime?: number;
  source?: CogniteRelationshipAsset;
  target?: CogniteRelationshipAsset;
}
export interface RelationshipsFilter {
  dataSetIds?: {
    id: number;
    value?: string;
  }[];
  labels?: {
    containsAny: {
      externalId: string;
      value?: string;
    }[];
  };
  isActiveAtTime?: boolean;
  activeAtTime?: {
    max: number;
    min: number;
  };
  sourceExternalIds?: string[];
  targetTypes?: string[];
}
