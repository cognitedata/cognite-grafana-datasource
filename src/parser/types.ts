import { STSClientFilter } from './ts';

export const FilterTypes = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export type ParsedFilter = STSClientFilter;

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
