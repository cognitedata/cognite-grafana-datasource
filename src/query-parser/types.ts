export const FilterType = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';
export interface QueryFilter {
  filter: FilterType;
  value: string;
}

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export interface ParsedFilter extends QueryFilter {
  path: string;
}

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
