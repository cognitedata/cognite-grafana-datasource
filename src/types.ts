import {
  DataQueryOptions,
  DataQuery,
  TimeSeries,
  TimeRange as GrafanaTimeRange,
  DataSourceSettings,
} from '@grafana/ui';

export function isError(maybeError: DataQueryError | any): maybeError is DataQueryError {
  return (<DataQueryError>maybeError).error !== undefined;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 * @param name â€“ event name
 */
export const eventFactory = <T = undefined>(name: string): AppEvent<T> => {
  return { name };
};

export type QueryResponse = DataResponse<TimeSeries[]>;

export interface MetricDescription {
  readonly text: string;
  readonly value: number | string;
}

export type MetricFindQueryResponse = MetricDescription[];

export enum Tab {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Custom = 'Custom',
}

export enum ParseType {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Event = 'Event',
}

export interface TimeSeriesResponseItem {
  id: number;
  externalId?: string;
  name?: string;
  isString?: boolean;
  metadata?: object;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  description?: string;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export interface AssetQuery {
  target: string;
  includeSubtrees: boolean;
  old?: AssetQuery;
  timeseries?: TimeSeriesResponseItem[];
  func?: string;
  templatedTarget?: string;
}

export interface InputQueryTarget extends DataQuery {
  target: number | '';
  aggregation: string;
  granularity: string;
  error: string;
  label: string;
  tab: Tab;
  assetQuery: AssetQuery;
  expr: string;
  warning: string;
}

export interface QueryTarget extends InputQueryTarget {
  target: number;
}

export type QueryFormat = 'json';

export type QueryOptions = DataQueryOptions<InputQueryTarget>;

export type Tuple<T> = [T, T];

export interface Range<T> {
  min?: T;
  max?: T;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 */
export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export interface DataSourceRequestOptions {
  url: string;
  method: HttpMethod;
  retry?: number;
  requestId?: string;
  headers?: { [s: string]: string };
  silent?: boolean;
  data?: object;
}

export interface Timestamp {
  timestamp: number;
}

export type MetaResponses = Responses<ResponseMetadata, DataQueryRequestResponse>;

export type SuccessResponse<Metadata, Response> = { metadata: Metadata; result: Response };
export type FailResponse<Metadata> = { metadata: Metadata; error: any };
export type Responses<Metadata, Response> = {
  failed: FailResponse<Metadata>[];
  succeded: SuccessResponse<Metadata, Response>[];
};

export interface TimeSeriesDatapoint extends Timestamp {
  value: string;
}

export interface TimeSeriesAggregateDatapoint extends Timestamp {
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export type Datapoints = Items<Datapoint>;

export interface DataQueryRequestResponse extends DataResponse<Datapoints> {
  config: {
    data: {
      aggregates: string;
      limit: number;
    };
  };
}

export interface RequestParams<DataType = any> {
  path: string;
  data: DataType;
  method: HttpMethod;
  params?: { [s: string]: any };
  requestId?: string;
  playground?: boolean;
  cacheTime?: string;
}

export type DataQueryError = {
  error: {
    cancelled?: boolean;
    data: {
      error?: {
        message: string;
        notFound?: string[];
      };
    };
    status: number;
  };
};

export type QueriesData = {
  items: DataQueryRequestItem[];
  target: QueryTarget;
}[];

export type ResponseMetadata = { labels: string[]; target: QueryTarget };

export interface DataQueryAlias {
  alias: string;
  id: number;
  aggregate?: string;
  granularity?: string;
}

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export type DataQueryRequestItem = {
  expression?: string;
  start?: string | number;
  end?: string | number;
  limit?: number;
  granularity?: string;
  aggregates?: string[];
  id?: number;
};

export type Aggregates = Pick<DataQueryRequest, 'aggregates'>;
export type Granularity = Pick<DataQueryRequest, 'granularity'>;

export interface DataQueryRequest {
  items: DataQueryRequestItem[];
  start?: string | number;
  end?: string | number;
  limit?: number;
  aggregates?: string[];
  granularity?: string;
}

export interface Annotation {
  datasource: string;
  enable: boolean;
  hide: boolean;
  iconColor: string;
  limit: number;
  name: string;
  query: string;
  filter: string;
  error: string;
  type: string;
  tags: string[];
}

export interface AnnotationQueryOptions {
  range: GrafanaTimeRange;
  rangeRaw: GrafanaTimeRange;
  annotation: Annotation;
  dashboard: number;
}

export interface AnnotationResponse {
  annotation: Annotation;
  title: string;
  time: number;
  timeEnd?: number;
  text: string;
  tags?: string[];
  isRegion?: boolean;
}

export interface AnnotationSearchQuery {
  description: string;
  type: string;
  subtype: string;
  minStartTime: number;
  maxStartTime: number;
  minEndTime: number;
  maxEndTime: number;
  minCreatedTime: number;
  maxCreatedTime: number;
  minLastUpdatedTime: number;
  maxLastUpdatedTime: number;
  // format is {"k1": "v1", "k2": "v2"}
  metadata: string;
  assetIds: number[];
  assetSubtrees: number[];
  sort: 'startTime' | 'endTime' | 'createdTime' | 'lastUpdatedTime';
  dir: 'asc' | 'desc';
  limit: number;
  offset: 0;
}

export interface Event {
  id: number;
  startTime: number;
  endTime: number;
  description: string;
  type: string;
  subtype: string;
  assetIds: number[];
  source: string;
  sourceId: string;
}

export type Events = Items<Event>;

export type DataEvents = DataResponse<Events>;

export type AnnotationQueryRequestResponse = DataResponse<DataEvents>;

export interface DataResponse<T> {
  data: T;
}

export type Items<T = object> = {
  items: T[];
};

export type CursorResponse<T> = DataResponse<Items<T> & { nextCursor?: string }>;

export type Response<T = object> = DataResponse<{
  items: T[];
}>;

export type TimeseriesFilterQuery = {
  filter?: {
    description?: string;
    assetSubtreeIds?: IdEither[];
    assetIds?: string[];
    [s: string]: any; // so we auto-support next features
  };
  cursor?: string;
} & Limit;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface VariableQueryData {
  query: string;
  error?: string;
}

export interface VariableQueryProps {
  query: string;
  onChange: (query: VariableQueryData) => void;
  datasource: any;
  templateSrv: any;
}

export interface CogniteDataSourceSettings extends DataSourceSettings {
  jsonData: {
    authType: string;
    defaultRegion: string;
    cogniteProject: string;
  };
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter: Filter;
}

export interface QueryRequestError {
  refId: string;
  error: string;
}

export interface QueryDatapointsLimitWarning {
  refId: string;
  warning: string;
}
