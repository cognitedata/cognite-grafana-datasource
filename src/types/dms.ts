// DMS API types
export interface DMSSpace {
  space: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSView {
  space: string;
  externalId: string;
  version: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSInstanceId {
  space: string;
  externalId: string;
}

export interface DMSFilter {
  and?: DMSFilter[];
  or?: DMSFilter[];
  not?: DMSFilter;
  equals?: {
    property: string[];
    value: any;
  };
  in?: {
    property: string[];
    values: any[];
  };
  range?: {
    property: string[];
    gte?: any;
    gt?: any;
    lte?: any;
    lt?: any;
  };
}

export interface DMSSearchRequest {
  view: {
    type: 'view';
    space: string;
    externalId: string;
    version: string;
  };
  query?: string;
  filter?: DMSFilter;
  limit?: number;
  properties?: string[];
}

export interface DMSSearchResponse {
  items: DMSInstance[];
}

export interface DMSInstance {
  instanceType: 'node' | 'edge';
  space: string;
  externalId: string;
  version: number;
  lastUpdatedTime: number;
  createdTime: number;
  deletedTime?: number;
  properties?: {
    [space: string]: {
      [view: string]: {
        [property: string]: any;
      };
    };
  };
}
