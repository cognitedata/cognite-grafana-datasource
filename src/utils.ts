import { isNil, omitBy, get } from 'lodash';
import { stringify } from 'query-string';
import ms from 'ms';
import { QueryOptions, QueryTarget } from './types';
import { FilterTypes, ParsedFilter } from './parser/types';

export function getQueryString(obj: any) {
  return stringify(omitBy(obj, isNil));
}

export function toGranularityWithLowerBound(milliseconds: number, lowerBound = 1000): string {
  return ms(Math.max(milliseconds, lowerBound));
}

// used for generating the options.requestId
export function getRequestId(options: QueryOptions, target: QueryTarget) {
  return `${options.dashboardId}_${options.panelId}_${target.refId}`;
}

export const applyFilters = <T>(objs: T[], filters: ParsedFilter[]): T[] => {
  if (!filters.length) {
    return objs;
  }

  return objs.filter((obj) => filters.every((filter) => checkFilter(obj, filter)));
};

export const checkFilter = <T>(obj: T, { path, filter, value }: ParsedFilter): boolean => {
  const valueToFilter = get(obj, path, null);
  const regex = new RegExp(`^${value}$`);

  switch (filter) {
    case FilterTypes.RegexEquals:
      return regex.test(valueToFilter);
    case FilterTypes.RegexNotEquals:
      return !regex.test(valueToFilter);
    case FilterTypes.NotEquals:
      return value !== valueToFilter;
    default:
      return false;
  }
};
