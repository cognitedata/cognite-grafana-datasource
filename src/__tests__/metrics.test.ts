import { cloneDeep } from 'lodash';
import { getMockedDataSource } from './utils';
import { VariableQueryData } from '../types';

const { ds, backendSrvMock } = getMockedDataSource();
const assetsResponse = {
  data: {
    items: [
      { id: 123, name: 'asset 1', description: 'test asset 1', metadata: { key1: 'value1' } },
      { id: 456, name: 'asset 2', description: 'test asset 2', metadata: { key1: 'value2' } },
      { id: 789, name: 'asset 3', description: 'test asset 3', metadata: { key1: 'value3' } },
      { id: 999, name: 'foo', description: 'bar', metadata: { key1: 'value1' } },
    ],
  },
};

describe('Metrics Query', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
    jest.clearAllMocks();
  });
  describe('Given an empty metrics query', () => {
    const variableQuery: VariableQueryData = {
      query: '',
    };
    it('should return empty array', async () => {
      const result = await ds.metricFindQuery(variableQuery);

      expect(result).toEqual([]);
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
    });
  });

  describe('Given a metrics query with no options', () => {
    let result;
    const variableQuery: VariableQueryData = {
      query: 'assets{}',
    };
    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(assetsResponse));
      result = await ds.metricFindQuery(variableQuery);
    });

    afterAll(() => {
      jest.clearAllMocks();
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return all assets', () => {
      expect(result.length).toEqual(assetsResponse.data.items.length);
    });
  });

  describe('Given a simple metrics query', () => {
    let result;
    const id = 123;
    const variableQuery: VariableQueryData = {
      query: `assets{id=${id}}`,
    };
    const response = cloneDeep(assetsResponse);
    response.data.items = assetsResponse.data.items.filter(item => item.id === id);

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(response));
      result = await ds.metricFindQuery(variableQuery);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct assets', () => {
      const resultIds = result.map(({ value }) => value);

      expect(result.length).toEqual(1);
      expect(resultIds.includes(id)).toBeTruthy();
    });
  });

  describe('Given a metrics query with filters', () => {
    let result;
    const variableQuery: VariableQueryData = {
      query: "assets{description=~'test asset.*', metadata={key1!='value2', key1!~'.*3'}}",
    };

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(assetsResponse));
      result = await ds.metricFindQuery(variableQuery);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct assets', () => {
      const resultIds = result.map(({ value }) => value);

      expect(result.length).toEqual(1);
      expect(resultIds.includes(123)).toBeTruthy();
    });
  });

  describe('Given a metrics query with unmatched filters', () => {
    let result;
    const variableQuery: VariableQueryData = {
      query: "assets{description=~'non-matched filter.*', metadata={key1!='value2'}}",
    };

    beforeAll(async () => {
      backendSrvMock.datasourceRequest = jest
        .fn()
        .mockImplementation(() => Promise.resolve(assetsResponse));
      result = await ds.metricFindQuery(variableQuery);
    });

    it('should generate the correct request', () => {
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
      expect(backendSrvMock.datasourceRequest.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should return the correct assets', () => {
      expect(result.length).toEqual(0);
    });
  });

  describe('Given an incomplete metrics query', () => {
    const variableQuery: VariableQueryData = {
      query: 'asset',
    };
    it('should return an empty array', async () => {
      const result = await ds.metricFindQuery(variableQuery);
      expect(result).toEqual([]);
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
    });
  });

  describe('Given an incorrect filter query', () => {
    it('should return an empty array if filter type is wrong', async () => {
      const variableQuery: VariableQueryData = {
        query: "assets{name~='foo'}",
      };
      const result = await ds.metricFindQuery(variableQuery);
      expect(result).toEqual([]);
      expect(backendSrvMock.datasourceRequest).not.toBeCalled();
    });
    it('should throw an error if filter regexp is wrong', async () => {
      const variableQuery: VariableQueryData = {
        query: "assets{name=~'*.foo'}",
      };
      expect(ds.metricFindQuery(variableQuery)).toThrowErrorMatchingSnapshot();
      expect(backendSrvMock.datasourceRequest).toBeCalledTimes(1);
    });
  });
});
