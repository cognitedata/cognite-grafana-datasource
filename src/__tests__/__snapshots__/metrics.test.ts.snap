// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics Query Given a metrics query with filters should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {},
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/assets/list",
}
`;

exports[`Metrics Query Given a metrics query with no options should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {},
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/assets/list",
}
`;

exports[`Metrics Query Given a metrics query with unmatched filters should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {},
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/assets/list",
}
`;

exports[`Metrics Query Given a nested variable query should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "parentIds": Array [
        123,
      ],
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/assets/list",
}
`;

exports[`Metrics Query Given a simple metrics query should generate the correct request 1`] = `
Object {
  "data": Object {
    "filter": Object {
      "id": 123,
    },
    "limit": 1000,
  },
  "method": "POST",
  "url": "/api/datasources/proxy/6/cdf-api-key/api/v1/projects/TestProject/assets/list",
}
`;

exports[`Metrics Query Given an incorrect filter query should throw an error if filter regexp is wrong 1`] = `[SyntaxError: Invalid regular expression: /^*.foo$/: Nothing to repeat]`;
